webapp/channels/src/selectors/calls.ts

```diff
@@ -2,16 +2,36 @@
 // See LICENSE.txt for license information.
 
 import {GlobalState} from 'types/store';
-import {suitePluginIds} from 'utils/constants';
+import {PostTypes, suitePluginIds} from 'utils/constants';
 import semver from 'semver';
+import {Post} from '@mattermost/types/posts';
+import {Channel} from '@mattermost/types/channels';
+import {General} from 'mattermost-redux/constants';
 
 export function isCallsEnabled(state: GlobalState, minVersion = '0.4.2') {
     return Boolean(state.plugins.plugins[suitePluginIds.calls] &&
         semver.gte(state.plugins.plugins[suitePluginIds.calls].version || '0.0.0', minVersion));
 }
 
-export function isCallsRingingEnabled(state: GlobalState) {
+// isCallsRingingEnabledOnServer is the flag for the ringing/notification feature in calls
+export function isCallsRingingEnabledOnServer(state: GlobalState) {
     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
     // @ts-ignore
     return Boolean(state[`plugins-${suitePluginIds.calls}`]?.callsConfig?.EnableRinging);
 }
+
+function isDmGmChannel(channelType: Channel['type']) {
+    return channelType === General.DM_CHANNEL || channelType === General.GM_CHANNEL;
+}
+
+export function callsWillNotify(state: GlobalState, post: Post, channel: Channel): boolean {
+    // Calls will notify if:
+    //  1. it's a custom_calls post (call has started)
+    //  2. in a DM or GM channel
+    //  3. calls is enabled and is v0.17.0+
+    //  4. calls ringing is enabled on the server
+    return post.type === PostTypes.CUSTOM_CALLS &&
+        isDmGmChannel(channel.type) &&
+        isCallsEnabled(state, '0.17.0') &&
+        isCallsRingingEnabledOnServer(state);
+}
```

Comment:
I am wondering if it would be feasible to keep this check on the plugin side to give us that little extra flexibility.

Reply:
I do like the idea of having more of this logic in Calls itself. Perhaps this would be a good opportunity to add a hook to let plugins intercept desktop notifications?

That feels like it would be a bit more future-proof rather than having to hardcode something into the web app for Calls in this case, and it could be used for other things like aggregating notifications or displaying them in some other way

Reply:
K, done. Thanks everyone, this is cleaner now. I rearranged the code a bit so that all params for `notifyMe` are sent to the hooks, in case they need to modify them. I defaulted to sending and returning all the args, in case a plugin in the future may need that.